src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,189 @@
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        T oldValue = getElement(index).value;
+        getElement(index).value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        return removeByIndex(index);
     }
 
     @Override
     public boolean remove(T object) {
-        return false;
+        int index = getIndex(object);
+        if (index == -1) {
+            return false;
+        }
+        checkIndex(index);
+        removeByIndex(index);
+        return true;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private class Node<T> {
+        private T value;
+        private Node prev;
+        private Node next;
+
+        public Node(Node prev, T value, Node next) {
+            this.prev = prev;
+            this.value = value;
+            this.next = next;
+        }
+    }
+
+    private void linkHead(T value) {
+        Node<T> h = head;
+        Node<T> newNode = new Node<>(null, value, h);
+        head = newNode;
+        if (h == null) {
+            tail = newNode;
+        } else {
+            h.prev = newNode;
+        }
+        size++;
+    }
+
+    private void linkTail(T value) {
+        Node<T> t = tail;
+        Node<T> newNode = new Node<>(t, value, null);
+        tail = newNode;
+        if (t == null) {
+            head = newNode;
+        } else {
+            t.next = newNode;
+        }
+        size++;
+    }
+
+    private void checkIndexElement(int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("The index out of bounds!");
+        }
+    }
+
+    private void checkIndex(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("The index out of bounds!");
+        }
+    }
+
+    private Node<T> getElement(int index) {
+        if (index < size >> 1) {
+            Node<T> current = head;
+            for (int i = 0; i < index; i++) {
+                current = current.next;
+            }
+            return current;
+        } else {
+            Node<T> current = tail;
+            for (int i = size - 1; i > index; i--) {
+                current = current.prev;
+            }
+            return current;
+        }
+    }
+
+    private int getIndex(T object) {
+        Node<T> current = head;
+        int index = -1;
+        for (int i = 0; i < size; i++) {
+            if (current == object || current.value == object
+                    || current.value != null && current.value.equals(object)) {
+                index = i;
+                break;
+            }
+            current = current.next;
+        }
+        return index;
+    }
+
+    private T removeByIndex(int index) {
```

Comment:
`removeByIndex(int index)` and `remove(int index) `- completely same meaning. Remove this method  and put this logic to `remove(int index)` method